web3.js/src/transaction.ts

```diff
@@ -221,11 +269,46 @@ export class Transaction {
    */
   _json?: TransactionJSON;
 
+  // Construct a transaction with a blockhash and lastValidBlockHeight
+  constructor(opts?: TransactionBlockhashCtor);
+
+  // Construct a transaction with a nonce
+  constructor(opts?: TransactionNonceCtor);
+
+  /**
+   * @deprecated `recentBlockhash` has been deprecated.
+   * Please pass `latestBlockhash: {blockHash: BlockHash, lastValidBlockHeight: number}` instead.
+   * This will be removed in a future version of @solana/web3.js.
+   */
+  constructor(opts?: TransactionCtorFields);
+
   /**
    * Construct an empty Transaction
    */
-  constructor(opts?: TransactionCtorFields) {
-    opts && Object.assign(this, opts);
+  constructor(
+    opts?:
+      | TransactionBlockhashCtor
+      | TransactionNonceCtor
+      | TransactionCtorFields,
+  ) {
+    if (!opts) {
+      return;
+      // eslint-disable-next-line no-prototype-builtins
+    } else if (opts.hasOwnProperty('recentBlockhash')) {
+      const oldOpts = opts as TransactionCtorFields;
+      Object.assign(this, oldOpts);
+      this.recentBlockhash = oldOpts.recentBlockhash;
+      // eslint-disable-next-line no-prototype-builtins
+    } else if (opts.hasOwnProperty('latestBlockhash')) {
+      const newOpts = opts as TransactionBlockhashCtor;
+      Object.assign(this, newOpts);
```

Comment:
This has the unfortunate side effect of setting `latestBlockhash` on `this`, whether you like it or not. Consider manually setting every property so that unintended consequences like that don't happen.